### PR TITLE
Append only state

### PR DESCRIPTION
**Pull request by Timo614**
## Description
The models (tokens, asset states) have their state appended to an append only table between checkpoints when the state is rolled up. This pull request adds a new append only table for each and a view that reflects the latest version of the data. In the future we should be able to handle the roll up logic by using the current additional_data_json from the view.

For replaying the data changes to nodes that need to catch up: they should be able to figure out which token is being used by using the asset ID (figured out via the asset state join -> digital asset) and the issue number. The data isn't denormalized to include either as they can be found via the joins.

## Motivation and Context
We don't want to allow changes to the state between checkpoints so this allows us to change the data without modifying it directly. These can be replayed to see the current state of the data.

## How Has This Been Tested?
I've added new tests and manually tested things a bit with the view.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by Timo614** [comment]
Based on our discussion today I pushed up a change to the append only state PR to instead include the entire state in each append only row. The view no longer merges the json for the append only state and instead is setup to just use the latest record as the state overriding the initial_data_json on the tokens or assets as needed. I removed the timestamp I was storing on the assets and tokens tables given the above.

File: node/migrations/V1588168331__create_tokens.sql [hunk 1]
```diff
@@ -28,4 +28,4 @@ EXECUTE PROCEDURE set_issue_number();
 CREATE INDEX index_tokens_uuid ON tokens (id);
 CREATE INDEX index_tokens_owner_pub_key ON tokens (owner_pub_key);
 CREATE INDEX index_tokens_status ON tokens (status);
-CREATE INDEX index_tokens_asset_state_id ON tokens (asset_state_id);
+CREATE UNIQUE INDEX index_tokens_asset_state_id_issue_number ON tokens (asset_state_id, issue_number);
```
Comment (dunnock) [inline node/migrations/V1588168331__create_tokens.sql]: I think issue_number uniqueness will fail under concurrency, but let's leave it to another PR

File: node/migrations/V1589211379__create_append_only_tables.sql [hunk 2]
```diff
@@ -0,0 +1,49 @@
+CREATE TABLE token_state_append_only (
+    id uuid PRIMARY KEY DEFAULT gen_random_uuid() NOT NULL,
+    token_id uuid NOT NULL references tokens(id),
```
Comment (dunnock) [inline node/migrations/V1589211379__create_append_only_tables.sql:3]: Should not we also make `status` (if we need to run consensus on token status) and possibly asset_state_id (as if token will change the asset_state?). Also perhaps asset_id as it is hash of fields metadata so it will change too.

File: node/migrations/V1589211379__create_append_only_tables.sql [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    FROM token_state_append_only
+    ORDER BY token_id, created_at DESC
+) tsao
+ON
+    t.id = tsao.token_id;
+
+CREATE TABLE asset_state_append_only (
+    id uuid PRIMARY KEY DEFAULT gen_random_uuid() NOT NULL,
+    asset_state_id uuid NOT NULL references asset_states(id),
+    state_data_json JSONB NOT NULL DEFAULT '{}',
```
Comment (dunnock) [inline node/migrations/V1589211379__create_append_only_tables.sql:32]: `asset_id` is going to change too as it is hash of metadata, also probably `status`?

File: node/migrations/V1589211379__create_append_only_tables.sql [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
+    updated_at TIMESTAMPTZ NOT NULL DEFAULT now()
+);
+CREATE INDEX index_asset_state_append_only_uuid ON asset_state_append_only (id);
+CREATE INDEX index_asset_state_append_only_asset_state_id_created_at ON asset_state_append_only (asset_state_id, created_at);
+
+CREATE OR REPLACE VIEW asset_states_view AS
```
Comment (dunnock) [inline node/migrations/V1589211379__create_append_only_tables.sql:39]: that's really cool

File: node/src/db/models/asset_states.rs [hunk 5]
```diff
@@ -20,11 +20,12 @@ pub struct AssetState {
     pub expiry_date: Option<DateTime<Utc>>,
     pub superseded_by: Option<uuid::Uuid>,
     pub initial_permission_bitflag: i64,
-    pub additional_data_json: Value,
+    pub initial_data_json: Value,
```
Comment (dunnock) [inline node/src/db/models/asset_states.rs:23]: so `initial_data_json` is the one which takes part in asset_id hashing as I got, and `additional_data_json` is not checkpointed yet as I got?

File: node/src/db/models/asset_states.rs [hunk 6]
```diff
@@ -134,13 +162,13 @@ mod test {
...
-        let mut additional_data_json = HashMap::new();
-        additional_data_json.insert("value", true);
+        let mut initial_data = HashMap::new();
+        initial_data.insert("value", true);
         let params = NewAssetState {
             name: "AssetName".to_string(),
             description: "Description".to_string(),
             asset_issuer_pub_key: PUBKEY.to_string(),
-            additional_data_json: serde_json::to_value(additional_data_json)?,
+            initial_data_json: serde_json::to_value(initial_data)?,
```
Comment (dunnock) [inline node/src/db/models/asset_states.rs]: could use `serde_json::json!()` macros:
```suggestion
        let params = NewAssetState {
            name: "AssetName".to_string(),
            description: "Description".to_string(),
            asset_issuer_pub_key: PUBKEY.to_string(),
            initial_data_json: serde_json::json!({"value": true}),
```

File: node/src/db/models/tokens.rs [hunk 7]
```diff
@@ -110,4 +134,59 @@ mod test {
...
+            },
+            &client,
+        )
+        .await?;
+        let mut expected_data = initial_data.clone();
+        expected_data.insert("value", Value::Null);
+        expected_data.insert("value2", json!(8));
+        expected_data.insert("value3", json!(2));
+        let token = Token::load(token.id, &client).await?;
+        assert_eq!(json!(expected_data), token.additional_data_json);
```
Comment (dunnock) [inline node/src/db/models/tokens.rs]: cool, just a bit less verbose way:
```suggestion
        let state_data_json = json!({"value": Value::Null, "value2": 8, "value3": 2});
        Token::store_append_only_state(
            NewTokenAppendOnly {
                token_id: token.id,
                state_data_json: state_data_json.clone(),
            },
            &client,
        )
        .await?;
        let token = Token::load(token.id, &client).await?;
        assert_eq!(state_data_json, token.additional_data_json);
```

**Review by dunnock — CHANGES_REQUESTED** [review]
That's great. Placed one suggestion to make tests less verbose and question on managing asset states

File: node/migrations/V1589211379__create_append_only_tables.sql (hunk 2 above)
Comment (dunnock) [reply node/migrations/V1589211379__create_append_only_tables.sql:3]: Asked this question in the group, let's discuss it perhaps

File: node/migrations/V1589211379__create_append_only_tables.sql (hunk 3 above)
Comment (dunnock) [reply node/migrations/V1589211379__create_append_only_tables.sql:32]: So we agreed that initial_state is the state from the beginning of times, not from the checkpoint, hence asset_id won't change. `status` might change though.

File: node/migrations/V1589211379__create_append_only_tables.sql (hunk 2 above)
Comment (dunnock) [reply node/migrations/V1589211379__create_append_only_tables.sql:3]: So we agreed that initial_state is the state from the beginning of times, not from the checkpoint, hence asset_id won't change. what about `status` though?

Comment (Timo614) [reply node/migrations/V1589211379__create_append_only_tables.sql:3]: Sounds good, will add that.

File: node/migrations/V1588168331__create_tokens.sql (hunk 1 above)
Comment (Timo614) [reply node/migrations/V1588168331__create_tokens.sql]: I've removed the unique keyword for now in this PR.

File: node/migrations/V1589211379__create_append_only_tables.sql (hunk 2 above)
Comment (Timo614) [reply node/migrations/V1589211379__create_append_only_tables.sql:3]: Added the status with the choices "Prepare, PreCommit, Commit". The view only shows the committed append only changes. I think we'll want to tie it back to the instruction record (especially if we're storing the instructions in the db) perhaps in the future but we can figure that out later in a subsequent PR if so.

File: node/src/db/models/asset_states.rs (hunk 5 above)
Comment (Timo614) [reply node/src/db/models/asset_states.rs:23]: We discussed this via a voice call earlier so just swinging back here:

The checkpoints are the hash of the state at the current append only period. We'll be leaving the initial data json alone unless the asset is migrated or perhaps a new command in the future to roll up the state into the initial state again to reduce the size of the asset (by condensing its state). The append only rows will normally be persisted for the sake of history. Should be able to verify the state of things by checking the hash of the append only data at the time of the checkpoint against what's stored in the checkpoint.

File: node/src/db/models/asset_states.rs (hunk 6 above)
Comment (Timo614) [reply node/src/db/models/asset_states.rs]: All set, I've updated all the instances of these.

File: node/migrations/V1589211379__create_append_only_tables.sql [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+SELECT
+    t.*,
+    COALESCE(tsao.state_data_json, t.initial_data_json) as additional_data_json
+FROM
+  tokens t
+LEFT JOIN
+(
+    SELECT DISTINCT ON(token_id) *
+    FROM token_state_append_only
+    WHERE status = 'Commit'
```
Comment (dunnock) [inline node/migrations/V1589211379__create_append_only_tables.sql:22]: That's it, cool! filtering out garbage

**Review by dunnock — APPROVED** [review]
LGTM!
